### ISSUE TITLE
Include D, and dub itself in build dependencies for reproducibility

### ISSUE THREAD

**Issue by biocyberman** [dlang/dub#1021]
Hi I am a Java `maven` and Scala `sbt` user. One thing I like about both tools is that it allows me to build with a specific Java/Scala version, and in case of sbt, a specific build tool version. This is very helpful for reproducibility. I think the same idea can be implemented for `dub`. For example:

- When I find an interesting project P on github that have not been updated for 2 years, but I want to try and develop it. The delopers of P run and tested it with `dub` and `D` versions at that time, and everything was fine. But now I try with newer version of dub and D, many things broken, even though I still can download other dependencies with correct versions. The solution is to have dub definition file (i.e. dub.json) to include `dub` and `D` versions at that time. The build with be based on those versions. And the build guarantied to be successful for practically longer period. 

This is also very helpful for production uses, in which developers are not pressed to update the project whenever there are new `D` and `dub` versions. 

That's my 2c of idea to make dub and D more appealing for adaption.

**Comment by p0nce** [comment]
Idea: DUB version could be something written in the generated `dub.selections.json` perhaps (an artifact that has to versionned if looking for reproducibility).

**Comment by wilzbach** [comment]
> This is also very helpful for production uses, in which developers are not pressed to update the project whenever there are new D

See also: https://github.com/dlang/dub/issues/863

**Comment by biocyberman** [comment]
@wilzbach Indeed, #863 issue is asking about something quite close to what this one is discussing. And to rephrase for clarity:

In essence, **A dub-backed project should be re-runnable from source**, in terms of time: after, say, 5 years; in terms of versions: with specified versions of dub, d, and direct dependencies. Dub should be able to pull everything by itself, not requiring developers to recreate the environment manually.

**Comment by p0nce** [comment]
A minimum front-end version, while a thing that does exist in all D project (sometimes unbeknownst to the author), would not help recompile projects.

An old project would fail to compile anyway because of:
- SemVer not respected
- added language warnings/errors
- things being deprecated then phased out (5 years is larger than the deprecation period), often for good reasons
- regressions

So yeah the only way to build for sure would have to have the **exact** same setup as the original (deps and compiler).

If you want to have a project build in 5 years, you would have to version `dub.selections.json`.
That seems like DVM should be improved instead so it parses some dub input (or output).
I mean, does `cargo` handle Rust version? Nope, it comes with the compiler.

A solution (to me) would be to have DVM parse `dub.selections.json` and in this file put the version of the compilers available (and the DUB version too). I think DUB knows this when it runs.

**Comment by dhasenan** [comment]
So, I think we can do this with two special packages:

* dlang: a virtual package representing the compiler. Its version reflects the language version as reported by the `__VERSION__` identifier.
* dub: dub, naturally.

A dub.sdl might look like:

```
name "foo"
dependency "dlang" version=">=2.081.0 <2.083.0"
dependency "dub" version="~>1.10.0"
```

If the current dub is within the requested version range, or there is no dependency on `dub`, it runs as normal. Otherwise it locates an appropriate dub with the standard dependency resolution techniques and runs it with the same arguments and environment.

The `dlang` package needs to be handled specially. We currently have three usable D compilers. Two of them lack certain DMDFE releases. So we add a preferred compiler to `settings.json` and create dub packages for dmd, ldc, and gdc. A `dlang` dependency is reinterpreted as a dependency on your preferred compiler (which can be any dub package). We add the `compiler` package type for them, just to make sure you don't do something silly like depend on a source library as your compiler.

You might also depend on a `dmd` compiler package, for instance, and that means your package must be compiled with dmd.

We save a `dlang` dependency in dub.selections.json as `dlang:<actual compiler package>` to distinguish it from a direct dependency on a compiler. We prefer the system compiler (if there is one) over a downloaded compiler.

If you don't depend on dlang or any compiler package, we just use the system compiler, and we don't record a compiler dependency in dub.selections.json.

This means that dub needs to know what compiler version it's dealing with. We'll need to record system compiler paths, dates modified, and BuildPlatform info in `~/.dub/compilers.json` or the like. When we detect changes, we'll rerun platform detection.

We'll only use the first compiler of each well-known family in $PATH as a system compiler.

----

Okay, that's a bit complex. But I think it covers everything, yes? And the dub and compiler bits are reasonably separable.

**Comment by ohdatboi** [comment]
I like the idea, but there's one _small_ problem: dmd's releases doesn't follow semver which means that `dependency "dlang" version=">=2.081.0 <2.083.0"` is not a valid version specification.

**Comment by dhasenan** [comment]
The language doesn't follow semver, but it does have a version number. That just means it would be unwise to depend on version range `~>2.081.0`.

**Comment by ohdatboi** [comment]
These version numbers doesn't follow semver and that's the problem. With semver we know when breaking changes are introduced but without it `2.081.2` can build an application just fine, but `2.082.0` might have breaking changes.

The only solution to that is to specify a version range, like `dependency "dlang" version=">=2.081.0 <2.083.0"` as you proposed but that would require developer to compile their package from time to time, checking if build fails and updating that specification. That kinda kills any purpose of such feature because with semver I could just specify `~>2.81` and _never_ update the package again - it would never fail to compile.

I think we could just work on such thing to check dub version (#1531) and when dmd finally adopts semver, do the same for it.

**Comment by dhasenan** [comment]
Depending on a specific dub and dlang version is primarily useful for executable targets, and they can pretty easily depend on an exact version. But it would be a lot better if dmd used semantic versioning.

**Comment by Geod24** [comment]
> But it would be a lot better if dmd used semantic versioning.

We're (slowly) getting there ;)

**Comment by biocyberman** [comment]
I have no hands-on experience with hacking dub and dlang. But from "end-user" point of view,  the thing I want to start with is like this (can be done with a Bash script, if specific versions of dependecies are really downloadable):

1. at the project folder, run some magic command (./buildthis.sh)
2. `buildthis.sh` read dub.json (or dub.selection.json, I don't know) and start fetching the correct version of `dlang` and `dub` to `./build` temp directory.
3. `buildthis.sh` call `./build/dlang/bin/dub` and hand over the build process
4. `./build/dlang/bin/dub` fetches dependencies to ./.dub and build the project with `./build/dlang/bin/dmd` (or ldc, or gdc for that matters).

What are the obstacles for these steps? Is that the versioning style one of the first things we have to solve?

**Comment by wilzbach** [comment]
@biocyberman: FYI you might like my bootstrap template (https://github.com/wilzbach/d-bootstrap). It does allow you to fix DMD and LDC to a specific version (for Linux at least).

**Comment by dhasenan** [comment]
> I think it is also important to point out that I am assuming that dlang and dub self-sufficient or self-contained. It means, they are future-proof.

Their external dependencies are:

        /lib64/ld-linux-x86-64.so.2 
        libc.so.6 
        libdl.so.2 
        libgcc_s.so.1 
        libm.so.6 
        libpthread.so.0 
        librt.so.1 
        libstdc++.so.6 
        linux-vdso.so.1

linux-vdso and ld-linux-x86-64 are provided by Linux. libgcc_s is provided by gcc. The rest are defined by Posix, so they should be pretty stable. I think.
